sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,564 @@
+    n,
+    dim,
+    n_clusters,
+    warp_size,
+    l2_cache_size,
+    window_length_multiple,
+    cluster_window_per_thread_group,
+    number_of_load_iter,
+):
+    r = warp_size * window_length_multiple
+    local_size = r * cluster_window_per_thread_group
+    h = number_of_load_iter * cluster_window_per_thread_group
+
+    n_cluster_groups = math.ceil(n_clusters / r)
+    global_size = (math.ceil(n / local_size)) * (local_size)
+    n_dim_windows = math.ceil(dim / h)
+    window_shm_shape = (h, r)
+
+    inf = float32(math.inf)
+    f32zero = float32(0.0)
+    f32one = float32(1.0)
+
+    nb_cluster_items = n_clusters * (dim + 1)
+    nb_cluster_bytes = 4 * nb_cluster_items
+    l2_cache_size_allocation = 0.9
+    nb_centroids_private_copies = int(
+        (l2_cache_size * l2_cache_size_allocation) // nb_cluster_bytes
+    )
+
+    @dpex.kernel
+    def fused_kernel_fixed_window(
+        X,
+        current_centroids,
+        centroids_half_l2_norm,
+        inertia,
+        centroids_private_copies,
+        centroid_counts_private_copies,
+    ):
+        global_thread = dpex.get_global_id(0)
+        local_thread = dpex.get_local_id(0)
+
+        window_shm = dpex.local.array(shape=window_shm_shape, dtype=float32)
+        local_centroids_half_l2_norm = dpex.local.array(shape=r, dtype=float32)
+        partial_scores = dpex.private.array(shape=r, dtype=float32)
+
+        first_centroid_global_idx = 0
+
+        min_idx = 0
+        min_score = inf
+
+        window_col = local_thread % r
+        window_row_offset = local_thread // r
+
+        for _0 in range(n_cluster_groups):
+
+            for i in range(r):
+                partial_scores[i] = f32zero
+
+            half_l2_norm_idx = first_centroid_global_idx + local_thread
+            if local_thread < r:
+                if half_l2_norm_idx < n_clusters:
+                    l2norm = centroids_half_l2_norm[half_l2_norm_idx]
+                else:
+                    l2norm = inf
+                local_centroids_half_l2_norm[local_thread] = l2norm
+
+            global_window_col = first_centroid_global_idx + window_col
+
+            first_dim_global_idx = 0
+
+            for _1 in range(n_dim_windows):
+                load_first_dim_local_idx = 0
+                for load_iter in range(number_of_load_iter):
+                    window_row = load_first_dim_local_idx + window_row_offset
+                    global_window_row = first_dim_global_idx + window_row
+
+                    if (global_window_row < dim) and (
+                        global_window_col < n_clusters
+                    ):
+                        item = current_centroids[
+                            global_window_row, global_window_col
+                        ]
+                    else:
+                        item = f32zero
+
+                    window_shm[window_row, window_col] = item
+
+                    load_first_dim_local_idx += cluster_window_per_thread_group
```

Comment:
```suggestion
            first_feature_idx = 0

            for _1 in range(feature_window_size):
                load_first_dim_local_idx = 0
                for load_iter in range(number_of_load_iter):
                    feature_window_offset = load_first_dim_local_idx + window_row_offset
                    feature_idx = first_feature_idx + feature_window_offset

                    if (feature_idx < dim) and (centroid_idx < n_clusters):
                        value = current_centroids[feature_idx, centroid_idx]
                    else:
                        value = f32zero

                    window_shm[feature_window_offset, centroid_window_offset] = value

                    load_first_dim_local_idx += cluster_window_per_thread_group
```

Not yet sure about `load_first_dim_local_idx` and `window_row_offset`...

Reply:
Maybe `window_shm` could renamed `centroids_window` instead?